Add JSON output support to Emboss text format

AaronWebster opened the pull request:
Add JSON output support to Emboss text format

AaronWebster committed:
Add JSON output support to Emboss text format.

reventlov commented on compiler/back_end/cpp/generated_code_templates:
```diff
@@ -1,17 +1,3 @@
-// Copyright 2019 Google LLC
```
Unless Google has change its policies in the last few months, this header should stay.

reventlov commented on compiler/back_end/cpp/generated_code_templates:
```diff
@@ -29,14 +15,14 @@
  */
 #ifndef ${header_guard}
 #define ${header_guard}
-#include <stdint.h>
```
Either delete if these are truly unnecessary, or uncomment if they're used in here.

reventlov commented on compiler/back_end/cpp/generated_code_templates:
```diff
@@ -29,14 +15,14 @@
+// #include <utility>
 
-#include "runtime/cpp/emboss_cpp_util.h"
+#include "third_party/emboss/runtime/cpp/emboss_cpp_util.h"
```
This is an internal-ism.  Copybara will adjust the path at import.

reventlov commented on compiler/back_end/cpp/generated_code_templates:
```diff
@@ -323,20 +322,31 @@ ${write_fields}
-            emboss_reserved_local_stream->Write(",");
+          if (!emboss_reserved_local_field_options.json() ||
+              emboss_reserved_local_wrote_field) {
+            emboss_reserved_local_stream->Write(" ");
```
You have moved the prefix `Write(" ")` so that it no longer occurs for the first element in a structure.  I don't see how the `SkippedStructureFieldOutput` is still passing.

Also: the `Write(" ")` here does *not* happen for JSON, but there are spaces between fields in the JSON goldens in the new tests.

Did I miss something, or is there version skew between the test and implementation?

reventlov commented on compiler/back_end/cpp/testcode/text_format_test.cc:
```diff
@@ -93,6 +93,89 @@ TEST(TextFormat, UpdateFromText) {
+      "{\"one_byte_enum\": \"ZERO\", \"seven_bit_uint\": 1, \"one_bit_flag\": "
+      "false, \"one_byte_uint\": 2, \"two_byte_uint\": 1027, "
+      "\"four_byte_uint\": 134678021, \"eight_byte_uint\": "
+      "1157159078456920585, \"uint8_array\": [17, 18, 19, 20, 21, 22, 23, 24, "
```
So: a very "fun" thing about JSON is that all numeric values are 64-bit *floating-point* values.  The `"eight_byte_uint"` value here will not be parsed correctly by many JSON parsers, depending how closely they hew to the JSON specification.

The typical workaround for this is to serialize integers as base 10 ASCII strings, either:

*   All integers are serialized as strings (this is what Emboss does internally for the serialized IR)
*   Integers whose range is larger than +/-2^53 are serialized as strings (so any 8-byte integer would be serialized as a string, no matter what its current value)
*   Integers whose actual value is outside of [-2^53,+2^53] are serialized as strings (so an 8-byte integer might be serialized as a string or a double, depending on its current value)

I don't think there is a single correct answer here, so it might just need another config value.

reventlov commented on compiler/back_end/cpp/generated_code_templates:
```diff
@@ -295,11 +287,18 @@ ${write_fields}
     (void)emboss_reserved_local_wrote_field;
     if (emboss_reserved_local_options.multiline()) {
+      if (emboss_reserved_local_wrote_field &&
+          emboss_reserved_local_options.json()) {
```
You can delete the `&& ...json()` here and then lines 348-351 entirely, I believe.

AaronWebster replied:
That would break non-JSON multiline output, suggestion?

reventlov commented on compiler/back_end/cpp/generated_code_templates:
```diff
@@ -323,20 +322,31 @@ ${write_fields}
         }
-        emboss_reserved_local_stream->Write("${field_name}: ");
+        if (emboss_reserved_local_field_options.json()) {
+          emboss_reserved_local_stream->Write("\"${field_name}\": ");
```
Most JSON serializers will omit the space between `:` and the field value when generating compact output.

reventlov commented on runtime/cpp/emboss_text_util.h:
```diff
@@ -80,12 +83,19 @@ class TextOutputOptions final {
   bool multiline() const { return multiline_; }
   bool digit_grouping() const { return digit_grouping_; }
   bool comments() const { return comments_; }
   ::std::uint8_t numeric_base() const { return numeric_base_; }
```
These three should probably return `... && !json_` instead of having `if (comments() && !json())` and similar logic scattered around the code.

AaronWebster committed:
Merge master into text-output-json

AaronWebster committed:
Address review feedback for JSON output support

- Restore license header in generated_code_templates
- Re-enable standard C++ includes that were incorrectly commented out
- Revert include path to internal format (Copybara adjusts at import)
- Add JsonLargeIntegerHandling option to serialize 64-bit integers as
  quoted strings to avoid precision loss in JSON parsers
- Remove space between colon and value in JSON compact output
- Make comments(), digit_grouping(), and numeric_base() accessors
  automatically return appropriate values when JSON mode is enabled,
  eliminating scattered json checks throughout the code
- Simplify WriteIntegerViewToTextStream, WriteEnumViewToTextStream,
  and WriteArrayToTextStream by leveraging the new accessor behavior
- Update tests to reflect compact JSON format (no spaces after colons)
- Update golden files for template changes

AaronWebster committed:
Add JSON output support to Emboss text format

Add a `json` mode to TextOutputOptions so any view can be serialized as
JSON, e.g. `WriteToString(view, TextOutputOptions().Json(true))`:

- Field names are quoted and values are comma-separated; comments are
  suppressed.  Compact output by default; multiline JSON places each
  field on its own indented line.
- Enums serialize as quoted names (or bare numbers when the value has no
  name); bool as true/false; NaN and Infinity as quoted strings.
- comments(), digit_grouping(), and numeric_base() automatically return
  JSON-appropriate values when json() is set, so callers don't have to
  special-case JSON throughout.
- Integers wider than 32 bits -- and the numeric form of unnamed enum
  values whose underlying type is wider than 32 bits -- are serialized as
  quoted strings by default (JsonLargeIntegerHandling::kLargeAsString),
  avoiding silent precision loss in JSON parsers that treat all numbers as
  IEEE-754 doubles.  JsonLargeIntegerHandling::kAsNumber opts back into raw
  numeric output for consumers that preserve full 64-bit integers.

Adds JsonTestEnum/JsonTestArrayStruct/JsonTestStruct/JsonEnumStruct/
JsonFloatStruct fixtures to testdata/text_format.emb and coverage in
text_format_test.cc and emboss_array_view_test.cc (compact and multiline
output, skipped fields, large-integer and unnamed-enum handling, and
NaN/Infinity).  Regenerates the C++ goldens.
